src/server/sql/reading/calculate_compare.sql

```diff
@@ -0,0 +1,106 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+	The compare chart needs three pieces of raw information for each meter:
+
+	1. How much was used during this time period?
+	2. How much was used during the previous time period?
+	3. How much was used up to this point in the previous time period
+ */
+
+/*
+This shouldn't ever be looking at more than a few weeks of data, so we don't need to deal with compression.
+ */
+CREATE FUNCTION compare_readings(
+	meter_ids INTEGER[],
+	current_period_start TIMESTAMP,
+	current_period_end TIMESTAMP,
+	period_duration INTERVAL
+)
+	RETURNS TABLE(meter_id INTEGER, current_use REAL, prev_use_total REAL, prev_use_for_current REAL)
+AS $$
+DECLARE
+	prev_period_start TIMESTAMP;
+	curr_period_duration INTERVAL; -- How far into the current period we are
+	prev_period_partial_end TIMESTAMP;
+BEGIN
+	curr_period_duration := current_period_end - current_period_start;
+	ASSERT curr_period_duration <= period_duration, 'Current period is too long';
+	prev_period_start := current_period_start - period_duration;
+	prev_period_partial_end := prev_period_start + curr_period_duration;
+
+	RETURN QUERY
```

Comment:
All of the following selects follow the same logic so I'll mention them all together. When selecting readings and checking that they are in the time range, would it not be better to just be looking for time range overlaps and just scale down the value of the reading if it is not fully within the range? This may be too many computations though so it is understandable not to have it but it does ignore some data and is an underestimate for both current and previous.

Reply:
Yeah, that's likely a good idea. We should probably have an sql function that handles the scaling since it happens in so many places. It's definitely not too much of a performance hit since we should only be looking at small intervals here (as I mention in the comment at the top). I'll take a longer look at it.

Reply:
Yeah, this logic happens in at least 3 places.